File: mixtourTests/core/MIXCoreBoard_privateTests.m
//
//  MIXCoreBoard_privateTests.m
//  mixtour
//
//  Created by Carsten Wenderdel on 2013-07-27.
//  Copyright (c) 2013 Carsten Wenderdel. All rights reserved.
//

#import <XCTest/XCTest.h>

#import "MixCoreBoard_private.h"


@interface MIXCoreBoard_privateTests : XCTestCase

@end

@implementation MIXCoreBoard_privateTests

- (void)testSetPiecesDirectly {
    
    struct MIXCoreBoard coreBoard;
    resetCoreBoard(&coreBoard);
    
    XCTAssertEqual(numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerWhite), (uint8_t)20u, @"");
    XCTAssertEqual(numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerBlack), (uint8_t)20u, @"");
    XCTAssertEqual(playerOnTurn(&coreBoard), MIXCorePlayerWhite, @"");
    
    MIXCoreSquare square = MIXCoreSquareMake(1, 1);
    setPiecesDirectly(&coreBoard, square, 1, MIXCorePlayerBlack);
    
    XCTAssertEqual(heightOfSquare(&coreBoard, square), (uint8_t)1, @"");
    XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square, 0), MIXCorePlayerBlack, @"");
    for (int i = 0; i < 5; i++) {
        for (int j = 0; j < 5; j++) {
            MIXCoreSquare testSquare = MIXCoreSquareMake(i, j);
            if (testSquare.column != square.column || testSquare.line != square.line) {
                XCTAssertEqual(heightOfSquare(&coreBoard, testSquare), (uint8_t)0, @"");
            }
        }
    }
    
    XCTAssertEqual(numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerWhite), (uint8_t)20u, @"");
    XCTAssertEqual(numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerBlack), (uint8_t)19u, @"");
    XCTAssertEqual(playerOnTurn(&coreBoard), MIXCorePlayerWhite, @"");
    
    MIXCoreSquare square2 = MIXCoreSquareMake(2, 2);
    setPiecesDirectly(&coreBoard, square2, 3, MIXCorePlayerBlack, MIXCorePlayerWhite, MIXCorePlayerBlack);
    
    XCTAssertEqual(heightOfSquare(&coreBoard, square2), (uint8_t)3, @"");
    XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 0), MIXCorePlayerBlack, @"");
    XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 1), MIXCorePlayerWhite, @"");
    XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 2), MIXCorePlayerBlack, @"");
    XCTAssertEqual(numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerWhite), (uint8_t)19u, @"");
    XCTAssertEqual(numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerBlack), (uint8_t)17u, @"");
    XCTAssertEqual(playerOnTurn(&coreBoard), MIXCorePlayerWhite, @"");
    
    setPiecesDirectly(&coreBoard, square2, 3, MIXCorePlayerBlack, MIXCorePlayerBlack, MIXCorePlayerWhite);
    
    XCTAssertEqual(heightOfSquare(&coreBoard, square2), (uint8_t)6, @"");
    XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 0), MIXCorePlayerWhite, @"");
    XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 1), MIXCorePlayerBlack, @"");
    XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 2), MIXCorePlayerBlack, @"");
    XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 3), MIXCorePlayerBlack, @"");
    XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 4), MIXCorePlayerWhite, @"");
    XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 5), MIXCorePlayerBlack, @"");
    XCTAssertEqual(numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerWhite), (uint8_t)18u, @"");
    XCTAssertEqual(numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerBlack), (uint8_t)15u, @"");
    XCTAssertEqual(playerOnTurn(&coreBoard), MIXCorePlayerWhite, @"");
    
    for (int i = 0; i < 5; i++) {
        for (int j = 0; j < 5; j++) {
            MIXCoreSquare testSquare = MIXCoreSquareMake(i, j);
            if ((testSquare.column != square.column || testSquare.line != square.line)
                    && (testSquare.column != square2.column || testSquare.line != square2.line)) {
                XCTAssertEqual(heightOfSquare(&coreBoard, testSquare), (uint8_t)0, @"");
            }
        }
    }
}


- (void)testSetTurnDirectly {
    
    struct MIXCoreBoard coreBoard;
    resetCoreBoard(&coreBoard);
    
    XCTAssertEqual(numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerWhite), (uint8_t)20u, @"");
    XCTAssertEqual(numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerBlack), (uint8_t)20u, @"");
    
    XCTAssertEqual(playerOnTurn(&coreBoard), MIXCorePlayerWhite, @"");
    setTurnDirectly(&coreBoard, MIXCorePlayerWhite);
    XCTAssertEqual(playerOnTurn(&coreBoard), MIXCorePlayerWhite, @"");
    
    // Usually white makes the first move.
    // Lets see whether everything is consistent when black starts
    setTurnDirectly(&coreBoard, MIXCorePlayerBlack);
    XCTAssertEqual(playerOnTurn(&coreBoard), MIXCorePlayerBlack, @"");
    
    setPiece(&coreBoard, MIXCoreSquareMake(0, 0));
    XCTAssertEqual(playerOnTurn(&coreBoard), MIXCorePlayerWhite, @"");
    XCTAssertEqual(numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerWhite), (uint8_t)20, @"");
    XCTAssertEqual(numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerBlack), (uint8_t)19, @"");
}


@end

File: mixtourTests/core/MIXCoreBoard_privateTests.swift
//
//  MIXCoreBoard_privateTests.swift
//  mixtour
//
//  Created by Wenderdel, Carsten on 30/12/14.
//  Copyright (c) 2014 Carsten Wenderdel. All rights reserved.
//

import XCTest
import Core
@testable import mixtour

class MIXCoreBoard_privateTests : XCTestCase {
    
    func testSetPiecesDirectly() {
        
        var coreBoard = MIXCoreBoard()
        resetCoreBoard(&coreBoard)
        
        XCTAssertEqual(Core.numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerWhite), UInt8(20))
        XCTAssertEqual(Core.numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerBlack), UInt8(20))
        XCTAssertEqual(playerOnTurn(&coreBoard).rawValue, MIXCorePlayerWhite.rawValue)
        
        let square = MIXCoreSquare(column:1, line:1)
        setPiecesDirectlyWithList(&coreBoard, square, 1, getVaList([MIXCorePlayerBlack.rawValue]))
        
        XCTAssertEqual(heightOfSquare(&coreBoard, square), UInt8(1))
        XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square, 0).rawValue, MIXCorePlayerBlack.rawValue)
        for i in 0..<5 {
            for j in 0..<5 {
                let testSquare = MIXCoreSquare(column:UInt8(i), line: UInt8(j))
                if (testSquare.column != square.column || testSquare.line != square.line) {
                    XCTAssertEqual(heightOfSquare(&coreBoard, testSquare), UInt8(0))
                }
            }
        }
        
        XCTAssertEqual(Core.numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerWhite), UInt8(20))
        XCTAssertEqual(Core.numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerBlack), UInt8(19))
        XCTAssertEqual(playerOnTurn(&coreBoard).rawValue, MIXCorePlayerWhite.rawValue)
        
        let square2 = MIXCoreSquare(column: 2, line: 2)
        Core.setPiecesDirectlyWithList(&coreBoard, square2, 3, getVaList([MIXCorePlayerBlack.rawValue, MIXCorePlayerWhite.rawValue, MIXCorePlayerBlack.rawValue]))
        
        XCTAssertEqual(heightOfSquare(&coreBoard, square2), UInt8(3))
        XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 0).rawValue, MIXCorePlayerBlack.rawValue)
        XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 1).rawValue, MIXCorePlayerWhite.rawValue)
        XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 2).rawValue, MIXCorePlayerBlack.rawValue)
        XCTAssertEqual(Core.numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerWhite), UInt8(19))
        XCTAssertEqual(Core.numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerBlack), UInt8(17))
        XCTAssertEqual(playerOnTurn(&coreBoard).rawValue, MIXCorePlayerWhite.rawValue)
        
        Core.setPiecesDirectlyWithList(&coreBoard, square2, 3, getVaList([MIXCorePlayerBlack.rawValue, MIXCorePlayerBlack.rawValue, MIXCorePlayerWhite.rawValue]))
        
        XCTAssertEqual(heightOfSquare(&coreBoard, square2), UInt8(6))
        XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 0).rawValue, MIXCorePlayerWhite.rawValue)
        XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 1).rawValue, MIXCorePlayerBlack.rawValue)
        XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 2).rawValue, MIXCorePlayerBlack.rawValue)
        XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 3).rawValue, MIXCorePlayerBlack.rawValue)
        XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 4).rawValue, MIXCorePlayerWhite.rawValue)
        XCTAssertEqual(colorOfSquareAtPosition(&coreBoard, square2, 5).rawValue, MIXCorePlayerBlack.rawValue)
        XCTAssertEqual(Core.numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerWhite), UInt8(18))
        XCTAssertEqual(Core.numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerBlack), UInt8(15))
        XCTAssertEqual(playerOnTurn(&coreBoard).rawValue, MIXCorePlayerWhite.rawValue)
        
        for i in 0..<5 {
            for j in 0..<5 {
                let testSquare = MIXCoreSquare(column: UInt8(i), line: UInt8(j))
                if ((testSquare.column != square.column || testSquare.line != square.line)
                    && (testSquare.column != square2.column || testSquare.line != square2.line)) {
                        XCTAssertEqual(heightOfSquare(&coreBoard, testSquare), UInt8(0))
                }
            }
        }
    }
    
    
    func testSetTurnDirectly() {
        
        var coreBoard = MIXCoreBoard()
        resetCoreBoard(&coreBoard)
        
        XCTAssertEqual(Core.numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerWhite), UInt8(20))
        XCTAssertEqual(Core.numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerBlack), UInt8(20))
        
        XCTAssertEqual(playerOnTurn(&coreBoard).rawValue, MIXCorePlayerWhite.rawValue)
        setTurnDirectly(&coreBoard, MIXCorePlayerWhite)
        XCTAssertEqual(playerOnTurn(&coreBoard).rawValue, MIXCorePlayerWhite.rawValue)
        
        // Usually white makes the first move.
        // Lets see whether everything is consistent when black starts
        setTurnDirectly(&coreBoard, MIXCorePlayerBlack)
        XCTAssertEqual(playerOnTurn(&coreBoard).rawValue, MIXCorePlayerBlack.rawValue)
        
        makeMove(&coreBoard, MIXCoreMoveMakeSet(MIXCoreSquare(column: 0, line: 0)))
        XCTAssertEqual(playerOnTurn(&coreBoard).rawValue, MIXCorePlayerWhite.rawValue)
        XCTAssertEqual(Core.numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerWhite), UInt8(20))
        XCTAssertEqual(Core.numberOfPiecesForPlayer(&coreBoard, MIXCorePlayerBlack), UInt8(19))
    }
    
}
